File: algorithms/0725. Split Linked List in Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k)
    {
        ListNode* c = head;
        int n = 0;
        
        while (c != NULL)
        {
            ++n;
            c = c->next;
        }
        
        int pn = n / k;
        int r = n % k;
        
        vector<ListNode*> ans(k, NULL);
        c = head;
        
        for (int i = 0; i < k; ++i)
        {
            ListNode* ch = c;
            
            for (int j = 0; j < pn + (i < r ? 1 : 0) - 1; ++j)
            {
                if (c != NULL)
                {
                    c = c->next;
                }
            }
            
            if (c != NULL)
            {
                ListNode* p = c;
                c = c->next;
                p->next = NULL;
            }
            
            ans[i] = ch;
        }
        
        return ans;
    }
};

File: 0725. Split Linked List in Parts.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode[] SplitListToParts(ListNode head, int k)
    {
        var result = new List<ListNode>();
        var length = GetLength(head);
        var avgSize = length / k;
        var remainder = length % k;
        var current = head;
        var previous = default(ListNode);

        for (int i = 0; i < k; ++i)
        {
            result.Add(current);
            
            for (var j = 0; j < avgSize; ++j)
            {
                previous = current;
                current = current?.next;
            }

            if (remainder > 0)
            {
                previous = current;
                current = current?.next;
                --remainder;
            }

            if (previous != null)
            {
                previous.next = null;
            }
        }

        return result.ToArray();
    }

    private int GetLength(ListNode head)
    {
        var length = 0;
        var current = head;

        while (current != null)
        {
            current = current.next;
            ++length;
        }

        return length;
    }
}